docs/writing-tests/making-a-testing-plan.md

```diff
@@ -0,0 +1,504 @@
+test.
+
+It's important to read these sections carefully, though, because the
+distinction between "mandatory" behavior and "optional" behavior can be quite
+nuanced. In this case, the optional behavior is never allowed if the document's
+type has changed. That makes for a valid conformance test, one that verifies
+browsers don't return the same result when the document's type changes.
+
+### Don't dive too deep
+
+Algorithms are usually composed of many other algorithms which themselves are
+defined in terms of still more algorithms. It can be intimidating to consider
+exhaustively testing one of those "nested" algorithms, especially when they are
+shared by many different APIs.
+
+In general, you should start by writing only "surface tests" for the nested
+algorithms. That means only verifying that they exhibit the basic behavior you
+are expecting.
+
+It's definitely important to test exhaustively, but it's just as important to
+do so in a structured way. Reach out to the test suite's maintainers to learn
+if and how they have already tested those algorithms. In many cases, it's
+acceptable to test them in just one place (and maybe through a different API
+entirely), and rely only on surface-level testing everywhere else. While it's
+always possible for more tests to uncover new bugs, the chances may be slim.
+The time we spend writing tests is highly valuable, so we have to be efficient!
+
+*Example* The following algorithm from [the DOM
+standard](https://dom.spec.whatwg.org/) powers
+[`document.querySelector`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector):
+
+> To **scope-match a selectors string** *selectors* against a *node*, run these
+> steps:
+>
+> 1. Let *s* be the result of [parse a
+>    selector](https://drafts.csswg.org/selectors-4/#parse-a-selector)
+>    *selectors*.
+>    [[SELECTORS4]](https://dom.spec.whatwg.org/#biblio-selectors4)
+> 2. If *s* is failure, then
+>    [throw](https://heycam.github.io/webidl/#dfn-throw) a
+>    "[`SyntaxError`](https://heycam.github.io/webidl/#syntaxerror)"
+>    [DOMException](https://heycam.github.io/webidl/#idl-DOMException).
+> 3. Return the result of [match a selector against a
+>    tree](https://drafts.csswg.org/selectors-4/#match-a-selector-against-a-tree)
+>    with *s* and *node*'s
+>    [root](https://dom.spec.whatwg.org/#concept-tree-root) using [scoping
+>    root](https://drafts.csswg.org/selectors-4/#scoping-root) *node*.
+>    [[SELECTORS4]](https://dom.spec.whatwg.org/#biblio-selectors4).
+
+As described earlier in this guide, we'd certainly want to test the branch
+regarding the parsing failure. However, there are many ways a string might fail
+to parse--should we verify them all in the tests for `document.querySelector`?
+What about `document.querySelectorAll`? Should we test them all there, too?
+
+The answers depend on the current state of the test suite: whether or not tests
```

Comment:
I would actually be stronger here myself and say that the answer is "no". There should be separate tests for 'parse a selector', and if there aren't then the right answer is to add those, not do the testing in the `document.querySelector` test.

Reply:
Those are both fair points. Ok, I am persuaded, there's more nuance here than I was allowing for :)